{{cookiecutter.package_name}}/# COOKIECUTTER{% if cookiecutter.docker_image == "y" %}Dockerfile# COOKIECUTTER{% endif %}

```diff
@@ -0,0 +1,34 @@
+ARG PYTHON_VERSION=3.12
+FROM python:${PYTHON_VERSION}-alpine AS builder
+
+WORKDIR /app
+
+RUN apk add --no-cache \
+    curl \
+    gcc \
+    musl-dev \
+    libffi-dev \
+    openssl-dev
+
+RUN curl -sSL https://raw.githubusercontent.com/pdm-project/pdm/main/install-pdm.py | python3 -
+
+ENV PATH="/root/.local/bin:$PATH"
+
+COPY . .
+
+RUN pdm export > /tmp/requirements.txt
+
+ARG PDM_BUILD_SCM_VERSION="0.0.1"
+RUN pdm build -d /tmp/wheelhouse
+
+RUN pip install --no-cache-dir -r /tmp/requirements.txt
+RUN pip install --no-cache-dir /tmp/wheelhouse/*.whl
+
+FROM python:${PYTHON_VERSION}-alpine
+
+COPY --from=builder /usr/local/lib/python${PYTHON_VERSION%.*}/site-packages /usr/local/lib/python${PYTHON_VERSION%.*}/site-packages
+COPY --from=builder /usr/local/bin /usr/local/bin
```

Comment:
there is no point of doing venv "container" inside already containerized system, it would also significantly increase the image size - which as mentioned in other comment - while not affecting typical apps much, for CLI is problematic

Reply:
I think right now we use venv in django cookiecutter, but AFAIK only reason for that is that we went and used pdm in different fashion than I did here, and it has no ability to work without venv other than using some rejected-PEP implementation. Also for django apps they are that heavy that we simply care much less about the extra megabytes added with venv.